### PR TITLE
Add image embedding and batch processing interfaces

### PR DESCRIPTION
**Pull request by jtcramer**
Combines PRs #29, #31, #32, #33, and #34 to avoid dependencies between PRs.

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -113,54 +139,230 @@ The embddings can be loaded from disk using numpy:
...
+    image_filepath_list = [image_filepath1, image_filepath2, image_filepath3]
+
+    # Saves embeddings to '/path/to/file1.npz', '/path/to/file2.npz', and '/path/to/file3.npz'
+    openl3.process_image_file(image_filepath_list, batch_size=32)
+
+
+Processing video files
+**********************
+
+Video files can also be processed to extract both audio and image embeddings:
```
Comment (justinsalamon) [inline docs/tutorial.rst]: Will the timestamps of the audio and image embeddings be the same? Aligned? We need to provide some info here.

File: docs/tutorial.rst [hunk 2]
```diff
@@ -113,54 +139,230 @@ The embddings can be loaded from disk using numpy:
...
+    # and image embedding to '/path/to/file_image_suffix.npz'
+    openl3.process_video_file(video_filepath, audio_model=audio_model, image_model=image_model,
+                              suffix='suffix')
+
+    # Save audio embedding to '/different/dir/file_audio_suffix.npz'
+    # and image embedding to '/different/dir/file_image_suffix.npz'
+    openl3.process_video_file(video_filepath, audio_model=audio_model, image_model=image_model,
+                              suffix='suffix', output_dir='/different/dir')
+
+Multiple files can be processed as well (with a specified batch size used for processing) as follows:
```
Comment (justinsalamon) [inline docs/tutorial.rst]: What does the batch size represent?

File: docs/tutorial.rst [hunk 3]
```diff
@@ -198,17 +400,66 @@ You can disable this centering as follows:
...
+
+.. code-block:: shell
+
+    $ openl3 image /path/to/file.png
+
+This will create an output file at ``/path/to/file.npz``. Functionality regarding
+specifying models, multiple input files, verbosity, output directories, and suffixes
+behave the same as with extracting audio embeddings.
+
+Similarly as with the audio embedding models, you can change the batch size used as the input to the
```
Comment (justinsalamon) [inline docs/tutorial.rst]: what does the batch size represent?

File: docs/tutorial.rst [hunk 4]
```diff
@@ -198,17 +400,66 @@ You can disable this centering as follows:
...
+Extracting image embeddings
+***************************
+
+To compute embeddings for a single image file via the command line run:
+
+.. code-block:: shell
+
+    $ openl3 image /path/to/file.png
+
+This will create an output file at ``/path/to/file.npz``. Functionality regarding
```
Comment (justinsalamon) [inline docs/tutorial.rst]: This section should be self sufficient (i.e. not rely on audio section)

File: docs/tutorial.rst [hunk 5]
```diff
@@ -198,17 +400,66 @@ You can disable this centering as follows:
...
+**********************
+
+To compute embeddings for a single image file via the command line run:
+
+.. code-block:: shell
+
+    $ openl3 video /path/to/file.mp4
+
+This will create output files at ``/path/to/file_audio.npz`` and ``/path/to/file_image.npz``
+for the audio and image embeddings, respectively. Functionality regarding
```
Comment (justinsalamon) [inline docs/tutorial.rst]: Again, section needs to be self sufficient.

File: docs/tutorial.rst [hunk 6]
```diff
@@ -198,17 +400,66 @@ You can disable this centering as follows:
...
+specifying models, multiple input files, verbosity, output directories, and suffixes
+behave the same as with extracting audio embeddings.
+
+You can change the batch size used as the input to the audio embedding and image embedding models
```
Comment (justinsalamon) [inline docs/tutorial.rst]: batch size represents?

File: openl3/core.py [hunk 7]
```diff
@@ -35,18 +38,87 @@ def _pad_audio(audio, frame_len, hop_len):
...
     ----------
-    audio : np.ndarray [shape=(N,) or (N,C)]
-        1D numpy array of audio data.
-    sr : int
-        Sampling rate, if not 48kHz will audio will be resampled.
+    audio : np.ndarray [shape=(N,) or (N,C)] or list[np.ndarray]
+        1D numpy array of audio data or list of audio arrays for multiple
+        inputs.
+    sr : int or list[int]
+        Sampling rate, or list of sampling rates. If not 48kHz will audio will
```
Comment (justinsalamon) [inline openl3/core.py]: typo: will audio will

File: openl3/core.py [hunk 8]
```diff
@@ -98,71 +167,418 @@ def get_embedding(audio, sr, model=None, input_repr="mel256",
...
 
-    if audio_len < frame_len:
-        warnings.warn('Duration of provided audio is shorter than window size (1 second). Audio will be padded.',
-                      OpenL3Warning)
+def process_audio_file(filepath, output_dir=None, suffix=None, model=None,
+                       input_repr="mel256", content_type="music",
+                       embedding_size=6144, center=True, hop_size=0.1,
+                       batch_size=32, overwrite=False, verbose=True):
+    """
+    Computes and saves L3 embedding for given audio file
```
Comment (justinsalamon) [inline openl3/core.py]: for a

File: openl3/core.py [hunk 9]
```diff
@@ -98,71 +167,418 @@ def get_embedding(audio, sr, model=None, input_repr="mel256",
...
+        Loaded model object. If a model is provided, then `input_repr`,
+        `content_type`, and `embedding_size` will be ignored.
+        If None is provided, the model will be loaded using
+        the provided values of `input_repr`, `content_type` and
+        `embedding_size`.
+    input_repr : "linear", "mel128", or "mel256"
+        Spectrogram representation used for model. Ignored if `model` is
+        a valid Keras model.
+    content_type : "music" or "env"
+        Type of content used to train embedding. Ignored if `model` is
```
Comment (justinsalamon) [inline openl3/core.py]: used to train the embedding model

File: openl3/core.py [hunk 10]
```diff
@@ -98,71 +167,418 @@ def get_embedding(audio, sr, model=None, input_repr="mel256",
...
+        Loaded model object. If a model is provided, then `input_repr`,
+        `content_type`, and `embedding_size` will be ignored.
+        If None is provided, the model will be loaded using
+        the provided values of `input_repr`, `content_type` and
+        `embedding_size`.
+    input_repr : "linear", "mel128", or "mel256"
+        Spectrogram representation used for to train audio part of audio-visual
+        correspondence model. Ignored if `model` is a valid Keras model.
+    content_type : "music" or "env"
+        Type of content used to train embedding. Ignored if `model` is
```
Comment (justinsalamon) [inline openl3/core.py]: used to train the embedding model

File: openl3/core.py [hunk 11]
```diff
@@ -98,71 +167,418 @@ def get_embedding(audio, sr, model=None, input_repr="mel256",
...
+        else:
+            return embedding_list
+
+
+def process_image_file(filepath, output_dir=None, suffix=None, model=None,
+                       input_repr="mel256", content_type="music",
+                       embedding_size=8192, batch_size=32,
+                       overwrite=False, verbose=True):
+    """
+    Computes and saves L3 embedding for given image file
```
Comment (justinsalamon) [inline openl3/core.py]: for a given

File: openl3/core.py [hunk 12]
```diff
@@ -181,45 +597,293 @@ def process_file(filepath, output_dir=None, suffix=None, model=None,
     content_type : "music" or "env"
         Type of content used to train embedding. Ignored if `model` is
         a valid Keras model.
-    embedding_size : 6144 or 512
+    embedding_size : 8192 or 512
```
Comment (justinsalamon) [inline openl3/core.py]: We should make it clear in the tutorial that the allowed embeddings sizes for audio and image are different

File: openl3/core.py [hunk 13]
```diff
@@ -181,45 +597,293 @@ def process_file(filepath, output_dir=None, suffix=None, model=None,
...
+                       audio_batch_size=32, image_batch_size=32,
+                       overwrite=False, verbose=True):
+    """
+    Computes and saves L3 audio and video frame embeddings for given video file
 
-    embedding, ts = get_embedding(audio, sr, model=model, input_repr=input_repr,
-                                  content_type=content_type,
-                                  embedding_size=embedding_size, center=center,
-                                  hop_size=hop_size, verbose=1 if verbose else 0)
+    Note that image embeddings are computed for every frame of the video. Also
```
Comment (justinsalamon) [inline openl3/core.py]: This explanation should be added to the tutorial

File: openl3/core.py [hunk 14]
```diff
@@ -181,45 +597,293 @@ def process_file(filepath, output_dir=None, suffix=None, model=None,
...
+    image_model : keras.models.Model or None
+        Loaded audio model object. If a model is provided, then `input_repr`,
+        `content_type`, and `embedding_size` will be ignored.
+        If None is provided, the model will be loaded using
+        the provided values of `input_repr`, `content_type` and
+        `embedding_size`.
+    input_repr : "linear", "mel128", or "mel256"
+        Spectrogram representation used for audio model. Ignored if `model` is
+        a valid Keras model.
+    content_type : "music" or "env"
```
Comment (justinsalamon) [inline openl3/core.py]: used to train the embedding model

File: openl3/models.py [hunk 15]
```diff
@@ -82,14 +88,69 @@ def load_embedding_model_path(input_repr, content_type):
...
+    """
+    Returns a model with the given characteristics. Loads the model
+    if the model has not been loaded yet.
+
+    Parameters
+    ----------
+    input_repr : "linear", "mel128", or "mel256"
+        Spectrogram representation used for audio model.
+    content_type : "music" or "env"
+        Type of content used to train embedding.
```
Comment (justinsalamon) [inline openl3/models.py]: to train the embedding model

File: tests/generate_openl3_regression_data.py [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+reg_image_emb_linear = np.load(reg_image_emb_linear_path)
+new_image_emb = np.load(new_image_emb_path)
+new_image_emb_linear = np.load(new_image_emb_linear_path)
+
+assert np.allclose(reg_audio_emb['timestamps'], new_audio_emb['timestamps'])
+assert np.allclose(reg_audio_emb['embedding'], new_audio_emb['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_image_emb['embedding'], new_image_emb['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_audio_emb_linear['timestamps'], new_audio_emb_linear['timestamps'])
+assert np.allclose(reg_audio_emb_linear['embedding'], new_audio_emb_linear['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_image_emb_linear['embedding'], new_image_emb_linear['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
```
Comment (justinsalamon) [inline tests/generate_openl3_regression_data.py]: Are regression data generation and regression tests combined into a single file here? Ideally we should have separate files for generating the original regression data (ideally never have to re-run these files), and then test files just for running the regression tests.

File: tests/generate_openl3_regression_data.py [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+
+
+assert np.allclose(reg_audio_emb['timestamps'], new_audio_emb['timestamps'])
+assert np.allclose(reg_image_emb['timestamps'], new_image_emb['timestamps'])
+assert np.allclose(reg_audio_emb['embedding'], new_audio_emb['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_image_emb['embedding'], new_image_emb['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_audio_emb_linear['timestamps'], new_audio_emb_linear['timestamps'])
+assert np.allclose(reg_image_emb_linear['timestamps'], new_image_emb_linear['timestamps'])
+assert np.allclose(reg_audio_emb_linear['embedding'], new_audio_emb_linear['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
+assert np.allclose(reg_image_emb_linear['embedding'], new_image_emb_linear['embedding'], rtol=1e-05, atol=1e-06, equal_nan=False)
```
Comment (justinsalamon) [inline tests/generate_openl3_regression_data.py]: See previous comment about separating out generation of regression data and regression tests

File: tests/test_cli.py [hunk 18]
```diff
@@ -15,6 +16,7 @@
 
 TEST_DIR = os.path.dirname(__file__)
 TEST_AUDIO_DIR = os.path.join(TEST_DIR, 'data', 'audio')
+TEST_IMAGE_DIR = os.path.join(TEST_DIR, 'data', 'image')
```
Comment (justinsalamon) [inline tests/test_cli.py]: Aren't we missing the same for video as well?

File: tests/test_cli.py [hunk 19]
```diff
@@ -25,10 +27,22 @@
...
 # Regression file paths
 TEST_REG_DIR = os.path.join(TEST_DIR, 'data', 'regression')
 REG_CHIRP_44K_PATH = os.path.join(TEST_REG_DIR, 'chirp_44k.npz')
 REG_CHIRP_44K_LINEAR_PATH = os.path.join(TEST_REG_DIR, 'chirp_44k_linear.npz')
+REG_DAISY_PATH = os.path.join(TEST_REG_DIR, 'daisy.npz')
+REG_DAISY_LINEAR_PATH = os.path.join(TEST_REG_DIR, 'daisy_linear.npz')
+REG_BENTO_AUDIO_PATH = os.path.join(TEST_REG_DIR, 'bento_audio.npz')
+REG_BENTO_AUDIO_LINEAR_PATH = os.path.join(TEST_REG_DIR, 'bento_audio_linear.npz')
+REG_BENTO_IMAGE_PATH = os.path.join(TEST_REG_DIR, 'bento_image.npz')
+REG_BENTO_IMAGE_LINEAR_PATH = os.path.join(TEST_REG_DIR, 'bento_image_linear.npz')
```
Comment (justinsalamon) [inline tests/test_cli.py]: Aren't we missing the same for video as well?

File: tests/test_cli.py [hunk 20]
```diff
@@ -117,72 +156,170 @@ def test_run(capsys):
...
+    image_outfile = os.path.join(tempdir, 'daisy.npz')
+    assert os.path.isfile(image_outfile)
 
     # regression test
-    data_reg = np.load(REG_CHIRP_44K_PATH)
-    data_out = np.load(outfile)
+    audio_data_reg = np.load(REG_CHIRP_44K_PATH)
+    audio_data_out = np.load(audio_outfile)
+    image_data_reg = np.load(REG_DAISY_PATH)
+    image_data_out = np.load(image_outfile)
```
Comment (justinsalamon) [inline tests/test_cli.py]: Aren't we missing the same for video as well?

File: tests/test_cli.py [hunk 21]
```diff
@@ -117,72 +156,170 @@ def test_run(capsys):
...
+    image_outfile = os.path.join(tempdir, 'daisy_linear.npz')
+    assert os.path.isfile(image_outfile)
 
     # regression test
-    data_reg = np.load(REG_CHIRP_44K_LINEAR_PATH)
-    data_out = np.load(outfile)
+    audio_data_reg = np.load(REG_CHIRP_44K_LINEAR_PATH)
+    audio_data_out = np.load(audio_outfile)
+    image_data_reg = np.load(REG_DAISY_LINEAR_PATH)
+    image_data_out = np.load(image_outfile)
```
Comment (justinsalamon) [inline tests/test_cli.py]: Aren't we missing the same for video as well?

File: tests/test_cli.py [hunk 22]
```diff
@@ -117,72 +156,170 @@ def test_run(capsys):
...
+    audio_outfile = os.path.join(tempdir, 'bento_audio.npz')
+    assert os.path.isfile(audio_outfile)
+    image_outfile = os.path.join(tempdir, 'bento_image.npz')
+    assert os.path.isfile(image_outfile)
+
+    # regression test
+    audio_data_reg = np.load(REG_BENTO_AUDIO_PATH)
+    audio_data_out = np.load(audio_outfile)
+    image_data_reg = np.load(REG_BENTO_IMAGE_PATH)
+    image_data_out = np.load(image_outfile)
```
Comment (justinsalamon) [inline tests/test_cli.py]: Aren't we missing the same for video as well?

File: tests/test_cli.py [hunk 23]
```diff
@@ -219,7 +356,7 @@ def test_script_main():
     assert sorted(data_out.files) == sorted(data_out.files) == sorted(
         ['embedding', 'timestamps'])
     assert np.allclose(data_out['timestamps'], data_reg['timestamps'],
-                       rtol=1e-05, atol=1e-06, equal_nan=False)
+                       rtol=1e-05, atol=1e-05, equal_nan=False)
     assert np.allclose(data_out['embedding'], data_reg['embedding'],
-                       rtol=1e-05, atol=1e-06, equal_nan=False)
+                       rtol=1e-05, atol=1e-05, equal_nan=False)
```
Comment (justinsalamon) [inline tests/test_cli.py]: Generally are we missing tests for video files for the CLI?

File: tests/test_core.py [hunk 24]
```diff
@@ -1,16 +1,20 @@
...
+from skimage.io import imread
+import openl3
+import openl3.models
 from openl3.openl3_exceptions import OpenL3Error
 from openl3.openl3_warnings import OpenL3Warning
 
 
 TEST_DIR = os.path.dirname(__file__)
 TEST_AUDIO_DIR = os.path.join(TEST_DIR, 'data', 'audio')
+TEST_IMAGE_DIR = os.path.join(TEST_DIR, 'data', 'image')
```
Comment (justinsalamon) [inline tests/test_core.py]: Generally are we missing tests for video files?

File: docs/tutorial.rst [hunk 25]
```diff
@@ -113,54 +139,230 @@ The embddings can be loaded from disk using numpy:
     data = np.load('/path/to/file.npz')
     emb, ts = data['embedding'], data['timestamps']
 
-As with ``get_embedding``, you can load the model manually and pass it to ``process_file`` to avoid loading the model multiple times:
+Multiple files can be processed as well (with a specified batch size used for processing) as follows:
```
[the remark this answers is not in the document]
Comment (jtcramer) [reply docs/tutorial.rst]: This is actually the model input batch size (i.e. individual frames/windows)

File: tests/generate_openl3_regression_data.py (hunk 16 above)
Comment (jtcramer) [reply tests/generate_openl3_regression_data.py]: They aren't, it's just a sanity check to indicate if the output has changed when re-generating the regression data.

File: tests/test_cli.py (hunk 18 above)
Comment (jtcramer) [reply tests/test_cli.py]: Ah I included the video in the image folder, but I'll move it to a video specific folder.

File: tests/test_cli.py (hunk 22 above)
Comment (jtcramer) [reply tests/test_cli.py]: This is the video test

File: tests/test_cli.py (hunk 23 above)
Comment (jtcramer) [reply tests/test_cli.py]: No, every instance of "bento" refers to a video

**Comment by justinsalamon** [comment]
Oh I forgot - we need to bump the version for a new release!

**Pull request merged**